### PR TITLE
Fix named event scheduling/deleting

### PR DESCRIPTION
**Pull request by penrods**
While working on the Alarm skill I discovered several issues with the
event scheduler.  This PR cleans up those findings and resolves several
other potential issues:

1) To avoid thread synchronization issues, the EventScheduler had several
queues which independently held objects to be added/deleted/updated.  However, the order of the events was undefined and got mixed since they were all batched together.  So, for instance, if skill code performed:
   self.add_event("foo", self.handle_foo)
   if SomeReason:
       self.cancel_event("foo")
The actual order of queue handling would perform Remove first, then Add which resulted in "foo" not being found for delete, but then added and left as an active event.

Now the EventScheduler protects the list using a Lock and the queues have been removed.  Modifications to the list happen immediately after obtaining the lock and are not batched up.

2) One-time events were triggered while the event was still in the EventScheduler list.  Now the entry is removed before invoking the handler.

3) Within the MycroftSkill.add_event(name, handler) is a local 'wrapper' method that actually makes the callback.  The MycroftSkill.remove_event(name) method attempted to find entries in the events list and the associated handler entries in the self.emitter to remove.  However, the emitter actually held the wrapper(handler), not the handler itself.  So the emitter handlers were left behind.

This was a quiet bug until the next time you scheduled an event of the same name.  When that second event finally triggered, it would fire off both the new and the old handler -- which snowballed in the 'skill-alarm:Beep' case, doubling and redoubling with every beep.

Now this cancels all the emitter listeners by name.  There is a very slim chance that someone has registered a listener with the same name, but since it is namespaced to "skill-name:Event" I think this is pretty safe.

## Description
(Description of what the PR does, such as fixes # {issue number})

## How to test
The was discovered with a skill that had code like this:
```python
    def _play_beep(self):
        """ Play alarm sound file """
        next_beep = datetime.now() + timedelta(seconds=6)
        self.schedule_event(self._play_beep, next_beep, name='Beep')
        self.log.debug("Beeping for: "+str(next_beep))
```

## Contributor license agreement signed?
CLA [ X ] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

### CODE REVIEW

File: mycroft/skills/event_scheduler.py [hunk 1]
```diff
@@ -217,31 +198,35 @@ def get_event_handler(self, message):
...
         emitter_name = 'mycroft.event_status.callback.{}'.format(event_name)
         self.emitter.emit(message.reply(emitter_name, data=event))
 
     def store(self):
         """
             Write current schedule to disk.
         """
-        with open(self.schedule_file, 'w') as f:
-            json.dump(self.events, f)
+        with self.event_lock:
```
[the remark this answers is not in the document]
Comment (penrods) [reply mycroft/skills/event_scheduler.py]: Yeah, I debated about whether to use a 'with' here and also on the load.  In both cases, I don't think they do any technical good.  I was debating whether to do them for consistency and decided to wrap them in with's to be consistent.  I don't think there are significant technical costs to it.

But I can be convinced to eliminate them, too, if you think this is silly.

**Comment by penrods** [comment]
I didn't grok the Mock testing mechanism and call_args[0][0].  I restored pieces and updated tests appropriately.

File: mycroft/skills/event_scheduler.py [hunk 2]
```diff
@@ -14,7 +14,7 @@
 #
 import json
 import time
-from threading import Thread
+from threading import Thread, Lock
```
[the remark this answers is not in the document]
Comment (penrods) [reply mycroft/skills/event_scheduler.py]: Done!

File: mycroft/skills/event_scheduler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (forslund) [reply mycroft/skills/event_scheduler.py]: I'm good with this as it is.

**Comment by forslund** [comment]
I think this looks good. Seems to be an unrelated pep8 issue. Feel free to merge

**Comment by penrods** [comment]
The PEP8 warning was from me commenting out a long line in the French test, which pushed it over 80 characters.  I fixed and will wait for the all-green, then merge.

**Comment by penrods** [comment]
Grr, I don't know why Codacy is giving an error here, but it shows all clear when I dig in to it.  So I'm ignoring that one.

**Pull request merged**